File: src/sowolver.rs
use std::collections::HashMap;

#[repr(u8)]
#[derive(Debug, PartialEq, Eq)]
pub enum Ass {
    False,
    True,
    Unass
}

impl Ass {
    fn neg(&self) -> Ass {
        match self {
            Ass::False => Ass::True,
            Ass::True => Ass::False,
            Ass::Unass => Ass::Unass
        }
    }
}

#[derive(Debug)]
pub struct Var {
    pub ass: Ass,
    watched_true: Vec<usize>,
    watched_false: Vec<usize>,
    pub n: usize
}

impl Var {
    pub fn new(n: usize) -> Self {
        Self {
            ass: Ass::Unass,
            watched_true: Vec::new(),
            watched_false: Vec::new(),
            n
        }
    }
}

#[derive(Debug)]
pub struct Clause {
    lits: HashMap<usize, bool>,
    watched: [usize;2],
    n: usize
}

impl Clause {
    pub fn new(n: usize) -> Self {
        Self {
            lits: HashMap::new(),
            watched: [0, 0],
            n
        }
    }
}

pub struct SOwOlver {
    pub num_vars: usize,
    pub num_clauses: usize,
    pub vars: Vec<Var>,
    pub clauses: Vec<Clause>,
    ass_queue: Vec<(usize, bool)>
}

impl SOwOlver {
    pub fn signature(&self) -> &'static str {
        "Version 1"
    }

    pub fn new(num_vars: usize, num_clauses: usize) -> Self {
        let mut s = Self {
            num_vars,
            num_clauses,
            vars: Vec::with_capacity(num_vars),
            clauses: Vec::with_capacity(num_clauses),
            ass_queue: Vec::new()
        };
        for i in 0..num_vars {
            s.vars.push(Var::new(i + 1));
        }
        s.clauses.push(Clause::new(s.clauses.len() + 1));
        s
    }

    pub fn add(&mut self, var: usize, sign: bool) {
        if var == 0 {
            if self.clauses.last().unwrap().lits.len() == 1 {
                let var = self.clauses.last().unwrap().lits.keys().last().unwrap();
                self.ass_queue.push((*var, *self.clauses.last().unwrap().lits.get(var).unwrap()));
            }
            if self.clauses.len() < self.num_clauses {
                self.clauses.push(Clause::new(self.clauses.len() + 1));
            }
        }
        else {
            let clause = self.clauses.last_mut().unwrap();
            clause.lits.insert(var, sign);
            let v = self.vars.get_mut(var - 1).unwrap();
            let mut watched = false;
            if clause.watched[0] == 0 {
                clause.watched[0] = var;
                watched = true;
            }
            else if clause.watched[1] == 0 {
                clause.watched[1] = var;
                watched = true;
            }
            if watched {
                (*(if sign { &mut v.watched_true } else { &mut v.watched_false })).push(clause.n);
            }
        }
    }

    pub fn solve(&mut self) -> bool {
        self.dpll()
    }

    pub fn failed(&mut self, lit: Var) -> bool {
        todo!()
    }

    pub fn dpll(&mut self) -> bool {
        self.bcp();
        let decision = self.dlis();
        if decision == 0 { return true }
        self.ass_queue.push((decision, false));
        if self.dpll() {
            return true;
        }
        self.ass_queue.push((decision, true));
        self.dpll()
    }

    fn bcp(&mut self) -> bool {
        println!("Assignment Queue: {:?}", self.ass_queue);
        while let Some(pending_ass) = self.ass_queue.pop() {
            println!("Assigning {} {}", pending_ass.0, pending_ass.1);
            let var: &mut Var = self.vars.get_mut(pending_ass.0 - 1).unwrap();
            if pending_ass.1 && var.ass == Ass::False || !pending_ass.1 && var.ass == Ass::True { return false; }
            var.ass = if pending_ass.1 { Ass::True } else { Ass::False };

            let var = self.vars.get(pending_ass.0 - 1).unwrap();
            let clauses = if pending_ass.1 { &var.watched_false } else { &var.watched_true };
            for clause in clauses {
                let clause = self.clauses.get_mut(clause - 1).unwrap();
                if clause.watched[0] != pending_ass.0 {
                    if clause.watched[1] == pending_ass.0 {
                        clause.watched.swap(0, 1);
                    }
                    else {
                        continue;
                    }
                }
                let mut lit_found = false;
                for lit in &clause.lits {
                    if *lit.0 != clause.watched[1] && (self.vars.get(*lit.0 - 1).unwrap().ass == Ass::Unass ||
                        (*lit.1 && self.vars.get(lit.0 - 1).unwrap().ass == Ass::True || 
                        !*lit.1 && self.vars.get(lit.0 - 1).unwrap().ass == Ass::False)) {
                        println!("Relinking watch pointer in clause {}: {} -> {}", clause.n, clause.watched[0], *lit.0);
                        clause.watched[0] = *lit.0;
                        lit_found = true;
                        break;
                    }
                }
                if !lit_found {
                    clause.watched[0] = 0;
                    if clause.watched[1] == 0 {
                        return false;
                    }
                    else {
                        let ass = *clause.lits.get(&clause.watched[1]).unwrap();
                        println!("Unit Clause detected: {}, queueing {} {}", clause.n, clause.watched[1], ass);
                        self.ass_queue.push((clause.watched[1], ass));
                    }
                }
            }
        }
        true
    }

    fn dlis(&self) -> usize {
        for var in &self.vars {
            if var.ass == Ass::Unass {
                println!("Deciding {}", var.n);
                return var.n;
            }
        }
        0
    }
}

File: src/SOwOlver.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;


public class SOwOlver {

    Var[] vars;
    Clause[] clauses;
    List<Clause> learnedClauses;
    Queue<Lit> assQueue;
    IG ig;
    Clause curClause;

    public SOwOlver(int numClauses, int numVars) {
        clauses = new Clause[numClauses];
        vars = new Var[numVars];
        for (int i = 0; i < numVars; i++) {
            vars[i] = new Var(i+1);
        }
        learnedClauses = new ArrayList<>();
        assQueue = new LinkedList<>();
        ig = new IG();
        if (numClauses > 0) {
            clauses[0] = new Clause(1);
            curClause = clauses[0];
        }
    }

    public void add(int n) {
        add(getLit(n));
    }

    public void add(Lit lit) {
        if (curClause == null) {
            System.out.println("Clauses full, literal ignored");
            return;
        }
        if (lit == null) {
            if (curClause.lits.size() == 1) {
                Lit toAssign = curClause.lits.get(0);
                System.out.println("Unit clause " + curClause.n + ": Queueing " + toAssign);
                assQueue.add(toAssign);
                ig.imply(toAssign, curClause);
            }
            int n = curClause.n;
            if (n < clauses.length) {
                clauses[n] = new Clause(n+1);
                curClause = clauses[n];
            }
            else {
                curClause = null;
            }
        }
        else {
            curClause.lits.add(lit);
            if (curClause.watched[0] == null) {
                curClause.watched[0] = lit;
                if (lit.positive) {
                    lit.var.watchedTrue.add(curClause);
                }
                else {
                    lit.var.watchedFalse.add(curClause);
                }
            }
            else if (curClause.watched[1] == null) {
                curClause.watched[1] = lit;
                if (lit.positive) {
                    lit.var.watchedTrue.add(curClause);
                }
                else {
                    lit.var.watchedFalse.add(curClause);
                }
            }
        }
    }

    public boolean solve() {
        return dpll();
    }

    private boolean dpll() {
        boolean result = bcp();
        if (!result) {
            List<Var> vars = ig.conflict();
            System.out.println("Conflict, unassigning " + vars);
            for (Var var : vars) {
                var.ass = Var.Ass.Unass;
            }
            assQueue.clear();
            return false;
        }
        Lit decision = dlis();
        System.out.println("DLIS returned: "+decision);
        if (decision == null) {
            return true;
        }
        ig.decide(decision);
        assQueue.add(decision);
        if (dpll()) {
            return true;
        }
        ig.decide(decision);
        assQueue.add(decision);
        return dpll();
    }

    private boolean bcp() {
        boolean conflict = false;
        outer:
        while (!assQueue.isEmpty()) {
            Lit lit = assQueue.remove();
            System.out.println("Assigning " + lit);
            if (lit.var.ass != Var.Ass.Unass) {
                if (lit.var.ass != Var.Ass.from(lit.positive)) {
                    return false;
                }
                continue;
            }
            lit.var.ass = Var.Ass.from(lit.positive);

            List<Clause> clauses = lit.positive ? lit.var.watchedFalse : lit.var.watchedTrue;
            List<Clause> toRemove = new ArrayList<>();
            for (Clause clause : clauses) {
                if (clause.watched[0].var != lit.var) {
                    if (clause.watched[1].var == lit.var) {
                        Lit swap = clause.watched[0];
                        clause.watched[0] = clause.watched[1];
                        clause.watched[1] = swap;
                    }
                    else {
                        continue;
                    }
                }
                boolean litFound = false;
                for (Lit lit_it : clause.lits) {
                    if (lit_it.var != lit.var &&  lit_it != clause.watched[1] &&
                            (lit_it.var.ass == Var.Ass.Unass || lit_it.var.ass == Var.Ass.from(lit_it.positive))) {
                        //System.out.println("Relinking in " + clause.n + ": " + clause.watched[0] + " -> " + lit_it);
                        clause.watched[0] = lit_it;

                        toRemove.add(clause);
                        if (lit_it.positive) {
                            lit_it.var.watchedTrue.add(clause);
                        }
                        else {
                            lit_it.var.watchedFalse.add(clause);
                        }
                        litFound = true;
                        break;
                    }
                }
                if (!litFound) {
                    if (clause.watched[1] == null) {
                        return false;
                    }
                    else if (clause.watched[1].ass() == Var.Ass.Unass) {
                        if (ig.imply(clause.watched[1], clause)) {
                            System.out.println("Unit clause " + clause.n + ": Queueing " + clause.watched[1]);
                            assQueue.add(clause.watched[1]);
                        }
                        else {
                            System.out.println("Unit clause " + clause.n + ": Conflict when queueing " + clause.watched[1]);
                            conflict = true;
                            break outer;
                        }
                    }
                }
            }
            for (Clause clause : toRemove) {
                clauses.remove(clause);
            }
        }
        return !conflict;
    }

    private Lit dlis() {
        int varCount = vars.length;
        int[] litImprovements = new int[varCount*2];
        for (Clause clause : clauses) {
            if(!clauseSAT(clause)){
                for (Lit lit :clause.lits) {
                    int index = lit.var.n -1; //zero based var nums
                    if(!lit.positive)
                        index += varCount;
                    litImprovements[index]++;
                }
            }
        }

        for (Clause clause : learnedClauses) {
            if(!clauseSAT(clause)){
                for (Lit lit :clause.lits) {
                    int index = lit.var.n -1; //zero based var nums
                    if(!lit.positive)
                        index += varCount;
                    litImprovements[index]++;
                }
            }
        }

        int maxAt = 0;
        for (int i = 0; i < litImprovements.length; i++) {
            maxAt = litImprovements[i] > litImprovements[maxAt] ? i : maxAt;
        }
        if(litImprovements[maxAt]==0) return null;
        int varNum =(maxAt % varCount)+1;
        Var decision = Arrays.stream(vars).filter(it-> it.n == varNum).findFirst().get();
        return (maxAt>=varCount) ?decision.negativeLit : decision.positiveLit; 
      
    }

    private boolean clauseSAT(Clause clause) {
        for(Lit lit : clause.lits){
            if(lit.ass() == Var.Ass.True){
                return true;
            }
        }
        return false;
    }

    public Lit getLit(int n) {
        if (n == 0) {
            return null;
        }
        boolean positive = n > 0;
        if (!positive) {
            n = -n;
        }
        if (n > vars.length) {
            return null;
        }
        return positive ? vars[n-1].positiveLit : vars[n-1].negativeLit;
    }

}
